cabal-install/Distribution/Client/CmdLegacy.hs

```diff
@@ -0,0 +1,67 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ViewPatterns #-}
+module Distribution.Client.CmdLegacy (legacyCmd) where
+
+import Prelude ()
+import Distribution.Client.Compat.Prelude
+
+import Distribution.Simple.Command
+import Distribution.Simple.Utils
+    ( warn )
+import Distribution.Verbosity 
+    ( normal )
+
+import qualified Data.Text as T
+
+-- Duplicated code, but makes things cleaner and lets me keep how this happens a dirty little secret.
+makeCmd :: CommandUI flags -> (flags -> [String] -> action) -> CommandSpec action
+makeCmd ui action = CommandSpec ui ((flip commandAddAction) action) NormalCommand
+
+deprecationNote :: String
+deprecationNote = 
+    "This command is a part of the legacy v1 style of cabal usage.\n\n" ++
+                          
+    "Please switch to using either the new project style or the legacy v1- aliases,\n" ++
+    "as new-style projects will become the default in the next version of\n" ++
+    "cabal-install."
+
+
+oldCommands :: [(T.Text, T.Text)]
+oldCommands = (\c -> (c, "v1-" `T.append` c)) <$> cmds
+    where 
+        cmds = 
+            [ "build"
+            , "configure"
+            , "repl"
+            , "freeze"
+            , "run"
+            , "test"
+            , "bench"
+            , "haddock"
+            , "exec"
+            , "update"
+            , "install"
+            ]
+
+legacyCmd :: CommandUI flags -> (flags -> [String] -> globals -> IO action) -> [CommandSpec (globals -> IO action)]
+legacyCmd origUi@CommandUI{..} action = [makeCmd legacyUi legacyAction, makeCmd aliasedUi action]
+    where
+        updateMsg msg = T.unpack $ foldr (\(old, new) str -> T.replace old new str) (T.pack msg) oldCommands
+
+        legacyAction flags extra globals = warn normal deprecationNote >> action flags extra globals
+
+        legacyUi = origUi 
+            { commandNotes = Just $ \pname -> case commandNotes of
+                Just notes -> notes pname ++ "\n" ++ deprecationNote 
+                Nothing -> deprecationNote
+            }
+        
+        aliasedUi = origUi
+            { commandName = updateMsg commandName
+            , commandUsage = updateMsg . commandUsage
+            , commandDescription = (updateMsg .) <$> commandDescription
+            , commandNotes = Just $ \pname -> case commandNotes of
+                Just notes -> updateMsg (notes pname) ++ "\n" ++ deprecationNote
```

Comment:
This is achieving too much:

```
cabal new-run cabal -- install --help

...

Examples:
  cabal v1-install                     Package in the current directory
  cabal v1-install foo                 Package from the hackage server
  cabal v1-install foo-1.0             Specific version of a package
  cabal v1-install 'foo < 2'           Constrained package version
  cabal v1-install v1-haddock --bindir=$HOME/hask-bin/ --datadir=$HOME/hask-data/
                                    Change v1-installation destination
```

I think you're going to have to put the `v1` in each legacy command's `commandNotes`, and then strip it out in the `legacyUI` (rather than adding it in the `aliasedUI` and keeping the `commandNotes` as they are).